src/NServiceBus.Core/Recoverability/Recoverability.cs

```diff
@@ -0,0 +1,170 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using Config;
+    using ConsistencyGuarantees;
+    using Features;
+    using Settings;
+    using Transports;
+
+    class Recoverability : Feature
+    {
+        public Recoverability()
+        {
+            EnableByDefault();
+            DependsOnOptionally<DelayedDeliveryFeature>();
+
+            Prerequisite(context => !context.Settings.GetOrDefault<bool>("Endpoint.SendOnly"),
+                "Message recoverability is only relevant for endpoints receiving messages.");
+            Defaults(settings =>
+            {
+                settings.SetDefault(SlrNumberOfRetries, DefaultSecondLevelRetryPolicy.DefaultNumberOfRetries);
+                settings.SetDefault(SlrTimeIncrease, DefaultSecondLevelRetryPolicy.DefaultTimeIncrease);
+
+                settings.SetDefault(FlrNumberOfRetries, 5);
+
+                settings.SetDefault(FailureInfoStorageCacheSizeKey, 1000);
+            });
+        }
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            var errorQueue = context.Settings.ErrorQueueAddress();
+            context.Settings.Get<QueueBindings>().BindSending(errorQueue);
+
+            var transportTransactionMode = context.Settings.GetRequiredTransactionModeForReceives();
+            var failureInfoStorage = new FailureInfoStorage(context.Settings.Get<int>(FailureInfoStorageCacheSizeKey));
+            var localAddress = context.Settings.LocalAddress();
+
+            context.Pipeline.Register("MoveFaultsToErrorQueue", b => new MoveFaultsToErrorQueueBehavior(
+                b.Build<CriticalError>(),
+                localAddress,
+                transportTransactionMode,
+                failureInfoStorage), "Moves failing messages to the configured error queue");  //context.Pipeline.Register(new MoveFaultsToErrorQueueBehavior.Registration(context.Settings.LocalAddress(), transportTransactionMode, failureInfoStorage));
+
+            context.Pipeline.Register("AddExceptionHeaders", new AddExceptionHeadersBehavior(), "Adds the exception headers to the message");
+            context.Pipeline.Register(new FaultToDispatchConnector(errorQueue), "Connector to dispatch faulted messages");
+
+            if (IsDelayedRetriesEnabled(context.Settings))
+            {
+                var retryPolicy = GetDelayedRetryPolicy(context.Settings);
+
+                context.Pipeline.Register("SecondLevelRetries", b => new SecondLevelRetriesBehavior(retryPolicy, localAddress, failureInfoStorage), "Performs second level retries");
+            }
+
+            if (IsImmediateRetriesEnabled(context.Settings))
+            {
+                var maxRetries = GetMaxRetries(context.Settings);
+                var retryPolicy = new FirstLevelRetryPolicy(maxRetries);
+
+                context.Pipeline.Register("FirstLevelRetries", b => new FirstLevelRetriesBehavior(failureInfoStorage, retryPolicy), "Performs first level retries");
```

Comment:
the slr behavior was registered with `DependencyLifecycle.InstancePerCall` but it seems that there was no real reason for that?